### PR TITLE
Fix readme telemetry

### PR DESCRIPTION
**Pull request by chenriksson**
Fixes bug from #827 where TrackPackageReadMeChangeEvent AI was logged under package push event name.

In addition, wanted to add tests to ensure that each API uses the correct event name. This required some refactoring to replace the static Telemetry class with an interface/wrapper for TelemetryClient.

### CODE REVIEW

File: src/NuGetGallery/OData/QueryAllowed/ODataQueryVerifier.cs [hunk 1]
```diff
@@ -124,12 +123,11 @@ public static ODataQueryFilter V1Search
             }
             catch (Exception ex)
             {
-                var telemetryProperties = new Dictionary<string, string>();
-                telemetryProperties.Add(TelemetryService.CallContext, $"{telemetryContext}:{nameof(AreODataOptionsAllowed)}");
-                telemetryProperties.Add(TelemetryService.IsEnabled, $"{isFeatureEnabled}");
-
                 // Log and do not throw
```
Comment (scottbommarito) [inline src/NuGetGallery/OData/QueryAllowed/ODataQueryVerifier.cs]: nit: this comment is not very useful

File: src/NuGetGallery/Services/ITelemetryService.cs [hunk 2]
```diff
@@ -18,6 +19,14 @@ public interface ITelemetryService
 
         void TrackVerifyPackageKeyEvent(string packageId, string packageVersion, User user, IIdentity identity, int statusCode);
 
+        /// <summary>
+        /// Create a trace for an exception (Application Insights->traces)
```
Comment (scottbommarito) [inline src/NuGetGallery/Services/ITelemetryService.cs]: Why are we creating traces for any exceptions? Why not just track them all as exceptions?

Comment (scottbommarito) [inline src/NuGetGallery/Services/ITelemetryService.cs]: nit: this `ITelemetryService` isn't inherently for Application Insights--there could be another implementation that logs to somewhere else. We probably shouldn't explicitly call out Application Insights here.

File: src/NuGetGallery/Services/TelemetryService.cs [hunk 3]
```diff
@@ -5,20 +5,24 @@
...
+            public const string PackageReadMeChanged = "PackageReadMeChanged";
+        }
 
-        // Event types
-        public const string ODataQueryFilterEvent = "ODataQueryFilter";
-        public const string PackagePushEvent = "PackagePush";
-        public const string CreatePackageVerificationKeyEvent = "CreatePackageVerificationKeyEvent";
-        public const string VerifyPackageKeyEvent = "VerifyPackageKeyEvent";
-        public const string PackageReadMeChangeEvent = "PackageReadMeChanged";
+        private IDiagnosticsSource _trace;
```
Comment (scottbommarito) [inline src/NuGetGallery/Services/TelemetryService.cs]: Can we rename this to `_diagnosticsSource`?

File: tests/NuGetGallery.Facts/Framework/Fakes.cs [hunk 4]
```diff
@@ -80,11 +80,11 @@ public Fakes()
             {
                 Id = "FakePackage",
                 Owners = new List<User> {Owner},
-                Packages = new List<Package>
-                {
-                    new Package {Version = "1.0"},
-                    new Package {Version = "2.0"}
-                }
+            };
+            Package.Packages = new List<Package>
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Framework/Fakes.cs]: nit: why didn't you leave this nested?

File: tests/NuGetGallery.Facts/Services/TelemetryServiceFacts.cs [hunk 5]
```diff
@@ -20,6 +25,167 @@ public void ThrowsIfDiagnosticsServiceIsNull()
...
+
+                // Assert
+                service.TelemetryClient.Verify(c => c.TrackEvent(eventName,
+                    It.IsAny<IDictionary<string, string>>(),
+                    It.IsAny<IDictionary<string, double>>()),
+                    Times.Once);
+            }
+            
+            [Fact]
+            public void TrackPackageReadMeChangeEventThrowsIfPackageIsNull()
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/TelemetryServiceFacts.cs]: Can we put tests for the same event in inner classes?

```
public class ThePackageReadMeChangeEvent
{
    [Fact]
    public void TrackPackageReadMeChangeEventThrowsIfPackageIsNull()
    {
        ...
    }
}
```

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 6]
```diff
@@ -39,7 +39,12 @@ public class DefaultDependenciesModule : Module
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1502:CyclomaticComplexity", Justification = "This code is more maintainable in the same function.")]
         protected override void Load(ContainerBuilder builder)
         {
-            var diagnosticsService = new DiagnosticsService();
+            var telemetryClient = TelemetryClientWrapper.Instance;
+            builder.RegisterInstance(telemetryClient)
+                .As<ITelemetryClient>()
```
Comment (joelverhagen) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: I'm not sure this guy should be a singleton.

From what I can tell online it looks like `TelemetryClient` inherit data from the request context:
https://stackoverflow.com/questions/34709278/using-static-instance-of-telemetryclient-with-application-insights#comment57259442_34709678

File: src/NuGetGallery/Services/ITelemetryService.cs (hunk 2 above)
Comment (chenriksson) [reply src/NuGetGallery/Services/ITelemetryService.cs]: I added trace for user safe exceptions in #809 to help with support requests, due to bug #4787.

File: tests/NuGetGallery.Facts/Framework/Fakes.cs (hunk 4 above)
Comment (chenriksson) [reply tests/NuGetGallery.Facts/Framework/Fakes.cs]: I needed to populate the reverse relationship (Package.PackageRegistration), but that can't be done within the PackageRegistration initialization.

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs (hunk 6 above)
Comment (skofman1) [reply src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: This was a Singleton before (see Telemetry.cs), so no change in behavior here.

---
In reply to: [147026697](https://github.com/NuGet/NuGetGallery/pull/4905#discussion_r147026697) [](ancestors = 147026697)

Comment (chenriksson) [reply src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: Ah, thanks! Will update.

File: src/NuGetGallery/Services/TelemetryClientWrapper.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    /// <summary>
+    /// Wrapper for the Application Insights TelemetryClient class.
+    /// </summary>
+    public class TelemetryClientWrapper : ITelemetryClient
+    {
+        internal static TelemetryClientWrapper Instance = new TelemetryClientWrapper();
+
+        private TelemetryClientWrapper()
+        {
+            UnderlyingClient = new TelemetryClient();
```
Comment (skofman1) [inline src/NuGetGallery/Services/TelemetryClientWrapper.cs]: Consider making this thread safe. Probably nothing will happen if TelemetryClient gets initialized a bunch of times, but still....

File: tests/NuGetGallery.Facts/Services/TelemetryServiceFacts.cs (hunk 5 above)
Comment (chenriksson) [reply tests/NuGetGallery.Facts/Services/TelemetryServiceFacts.cs]: All of the TrackXEvent APIs are thin wrappers around TrackEvent and share some of the same tests (e.g., TrackEventNames, TrackEventNamesIncludesAllEvents). This is why I chose to keep them together.

File: src/NuGetGallery/Services/TelemetryClientWrapper.cs (hunk 7 above)
Comment (chenriksson) [reply src/NuGetGallery/Services/TelemetryClientWrapper.cs]: Should be safe... TelemetryClient is initialized in ctor, so each wrapper would have a single client instance. The wrapper is also implemented as a singleton with private ctor and static instance. I did update it to use Lazy for deferred initialization, which should be thread safe.

**Comment by chenriksson** [comment]
@joelverhagen Removed all non-EF (schema) helpers from the entities.

Totally onboard with the separation of EF entities from helpers, which could help prevent accidental schema changes as well.  I believe EF schema generation originally used partials for this, but with code first maybe it makes sense for us to create extensions for entities instead?

We already have this problem with existing entities... maybe I should file an issue to refactor, so that this new pattern is adopted?

**Comment by joelverhagen** [comment]
> We already have this problem with existing entities... maybe I should file an issue to refactor, so that this new pattern is adopted?

Seems reasonable. One extreme would be mapping DAL entities to business logic types and baking all query patterns we need down into the DAL. Then we can have all of the business logic/convenience properties and methods that we want -- on the BL types. This is extreme and a big deviation from what we have today though. Perhaps a middle ground is writing all of these convenience methods/properties as methods of in our `*Service` classes.

**Comment by chenriksson** [comment]
@joelverhagen Oops, my last comment was intended for PR #4904. I've got too many tabs open :-)

**Pull request merged**